### PR TITLE
mem_cache: sync-free in-place out-of-window SWA frees

### PR DESCRIPTION
**Pull request by sshleifer**
Out-of-window SWA frees run `torch.unique`/boolean-mask ops whose data-dependent output shapes force a stream sync under the in-flight overlapped forward — one free can stall a full scheduler pass. On an internal RL rolling workload this was ~139s of a 1,200s run; the fix alone was worth ~+5% end-to-end output throughput (11,411 -> 12,013 tok/s).

The freed range in `free_swa_out_of_window_slots` is one request's page-aligned segment: page representatives are the stride slice `free_index[::page_size]`, and a first-time-freed range below the eviction frontier has every mapping live, so the `> 0` filter is skippable. `free_swa_segment_inplace` therefore uses only fixed-shape ops (plus a new fixed-shape `free_page_ids` tail on the paged allocator and a `free_v_pages_lazy` twin on the multi-ended allocator). Same call site and timing; bitwise-identical end state (unit-tested, 13 cases). Partial mappings, unaligned inputs, and in-group calls fall back to the existing `free_swa`.

A debug-mode reference check (`SGLANG_DEBUG_MEMORY_POOL`) validates the fixed-shape contract against `torch.unique` on CPU.

---
### CI States

Latest PR Test (Base): :x: [Run #33756419396](https://github.com/sgl-project/sglang/actions/runs/33756419396)
Latest PR Test (Extra): :x: [Run #33756418601](https://github.com/sgl-project/sglang/actions/runs/33756418601)
Latest PR Test (AMD ROCm 7.2): :x: [Run #33756419170](https://github.com/sgl-project/sglang/actions/runs/33756419170)

### CODE REVIEW

File: python/sglang/srt/mem_cache/allocator/swa.py [hunk 1]
```diff
@@ -96,6 +96,7 @@ def __init__(
         self.is_not_in_free_group = True
         self.free_group = []
         self.swa_free_group = []
+        self._swa_mapping_may_be_partial = False
```
Comment (sshleifer) [inline python/sglang/srt/mem_cache/allocator/swa.py:102]: `_swa_mapping_may_be_partial` becomes True once a tail-only swa allocation has installed zero mappings; from then on the fixed-shape frees fall back to the filtering path.

File: python/sglang/srt/mem_cache/allocator/swa.py [hunk 2]
```diff
@@ -281,6 +282,7 @@ def alloc_extend_swa_tail(
         )
         if swa_tail_len < extend_num_tokens:
             self.clear_full_to_swa_mapping(alloc_full_indices[:-swa_tail_len])
+            self._swa_mapping_may_be_partial = True
```
Comment (sshleifer) [inline python/sglang/srt/mem_cache/allocator/swa.py:289]: Allocated-but-unmapped slots now exist, so the in-place segment free must keep filtering — this flag routes it to the fallback (see `free_swa_segment_inplace`).

File: python/sglang/srt/mem_cache/allocator/swa.py [hunk 3]
```diff
@@ -368,6 +370,47 @@ def free_swa(self, free_index: torch.Tensor):
...
+        if ps == 1:
+            swa_slots = self.full_to_swa_index_mapping[free_index]
+            self.swa_attn_allocator.free(swa_slots)
+            self.clear_full_to_swa_mapping(free_index)
+            return
+
+        page_reps = free_index[::ps]
+        swa_reps = self.full_to_swa_index_mapping[page_reps]
+        if self.swa_attn_allocator.debug_mode:
+            ref_pages = torch.unique(free_index.cpu() // ps)
```
Comment (sshleifer) [inline python/sglang/srt/mem_cache/allocator/swa.py]: This debug block is a reference check of the fixed-shape contract (runs on CPU; debug mode only).

File: python/sglang/srt/mem_cache/allocator/swa.py [hunk 4]
```diff
@@ -368,6 +370,47 @@ def free_swa(self, free_index: torch.Tensor):
...
+            got_pages = torch.sort(page_reps.cpu() // ps)[0]
+            assert torch.equal(ref_pages, got_pages), "range is not whole pages"
+            assert bool(
+                (swa_reps.cpu() > 0).all()
+            ), "out-of-window range has unmapped slots"
+        swa_page_ids = torch.sort(swa_reps // ps).values
```
Comment (sshleifer) [inline python/sglang/srt/mem_cache/allocator/swa.py]: Requests own disjoint pages, so the page representatives are distinct; a fixed-shape sort therefore reproduces exactly the ordering `free()`'s `torch.unique` would produce.

File: python/sglang/srt/mem_cache/allocator/swa.py [hunk 5]
```diff
@@ -516,6 +560,15 @@ def free_swa(self, free_index: torch.Tensor):
         else:
             self.free_group.append(self._copy_for_free_group(free_index))
 
+    def free_swa_segment_inplace(self, free_index: torch.Tensor, *, start_pos: int):
```
Comment (sshleifer) [inline python/sglang/srt/mem_cache/allocator/swa.py:633]: PureSWA case: the mapping is identity (never zeroed) and page_size == 1, so the only data-dependent op in `free_swa` is its `> 0` filter — and an out-of-window range holds real slot values, so the filter can be dropped. In-group behavior is exactly `free_swa`'s.

File: python/sglang/srt/mem_cache/multi_ended_allocator.py [hunk 6]
```diff
@@ -1039,6 +1039,34 @@ def _free_lazy(self, free_index: torch.Tensor) -> None:
...
+        so it never forces a stream sync under an in-flight forward. LAZY
+        mode only — eager compaction walks freed ids on the CPU and cannot be
+        made sync-free without redesign."""
+        assert self.lazy_compaction
+        if free_v_pages.numel() == 0:
+            return
+        self._stats_n_free_lazy += 1
+        with record_function("MultiEndedAlloc.free_v_pages_lazy"):
+            freed_p_pages = self.virtual_to_physical[free_v_pages]
+            neg_one = torch.full(
```
Comment (sshleifer) [inline python/sglang/srt/mem_cache/multi_ended_allocator.py]: The -1 must be device-resident, not a python scalar: `index_put_` with a python scalar copies it host->device with a blocking memcpy — a stream sync, the very thing this method exists to avoid.

File: python/sglang/srt/mem_cache/multi_ended_allocator.py [hunk 7]
```diff
@@ -2136,6 +2164,7 @@ def __init__(
         # Empty (not None) for the leak checker.
         self.free_pages = torch.empty(0, dtype=torch.int64, device=device)
         self.release_pages = torch.empty(0, dtype=torch.int64, device=device)
+        self._swa_mapping_may_be_partial = False
```
Comment (sshleifer) [inline python/sglang/srt/mem_cache/multi_ended_allocator.py:3231]: The parent class's tail-only-swa alloc path (which installs unmapped slots) is unsupported in the unified allocator, so this flag stays False here. It is read by `free_swa_segment_inplace`.

**Review by sshleifer — COMMENTED** [review]
Moved the PR's inline rationale comments into the review annotations below, anchored to the lines they described, so the diff itself stays comment-free.

-Robot

File: python/sglang/srt/mem_cache/multi_ended_allocator.py [hunk 8]
```diff
@@ -2478,6 +2507,33 @@ def free_swa(self, free_index: torch.Tensor) -> None:
...
+            not sa.lazy_compaction
+            or self._swa_mapping_may_be_partial
+            or start_pos % ps != 0
+            or n % ps != 0
+        ):
+            self.free_swa(free_index)
+            return
+        v = free_index.detach().to(torch.int64)
+        if ps == 1:
+            free_v_pages = v
```
Comment (sshleifer) [inline python/sglang/srt/mem_cache/multi_ended_allocator.py]: Raw order on purpose: at ps==1 `_free_lazy` skips `torch.unique`, so passing the tokens in request order gives bitwise parity.

File: test/registered/unit/mem_cache/test_sync_free_swa_inplace.py [hunk 9]
```diff
@@ -0,0 +1,387 @@
...
+        for i, seg in enumerate(segs):
+            start_pos = (
+                start_positions[i] if start_positions is not None else 0
+            )
+            if use_inplace:
+                alloc.free_swa_segment_inplace(seg, start_pos=start_pos)
+            else:
+                alloc.free_swa(seg)
+        if mixed_seg is not None:
+            alloc.free_swa(mixed_seg)
```
Comment (sshleifer) [inline test/registered/unit/mem_cache/test_sync_free_swa_inplace.py:130]: The mixed_seg free models a finished-request-style eager `free_swa` interleaved with the in-place frees.

File: test/registered/unit/mem_cache/test_sync_free_swa_inplace.py [hunk 10]
```diff
@@ -0,0 +1,387 @@
...
+
+    def test_mixed_group_equals_eager(self):
+        segs = _segments(seed=11, n_segments=4)
+        mixed = segs[-1]
+        segs = segs[:-1]
+        a = self._run(segs, use_inplace=False, group=True, mixed_seg=mixed)
+        b = self._run(segs, use_inplace=True, group=True, mixed_seg=mixed)
+        _assert_states_equal(self, a, b)
+
+    def test_partial_mapping_flag_falls_back(self):
```
Comment (sshleifer) [inline test/registered/unit/mem_cache/test_sync_free_swa_inplace.py:155]: Scenario: tail-only-swa allocations leave zero mappings; the flag routes the in-place call to the filtering `free_swa`, so states stay identical.

File: test/registered/unit/mem_cache/test_sync_free_swa_inplace.py [hunk 11]
```diff
@@ -0,0 +1,387 @@
...
+        segs = _segments(seed=5, n_segments=3)
+        a = self._run(segs, use_inplace=False, mapped_fraction=0.5, partial_flag=True)
+        b = self._run(segs, use_inplace=True, mapped_fraction=0.5, partial_flag=True)
+        _assert_states_equal(self, a, b)
+
+    def test_unaligned_start_falls_back(self):
```
Comment (sshleifer) [inline test/registered/unit/mem_cache/test_sync_free_swa_inplace.py:161]: Scenario: start_pos not page-aligned -> legacy `free_swa` (which expands to full pages); states identical.

File: test/registered/unit/mem_cache/test_sync_free_swa_inplace.py [hunk 12]
```diff
@@ -0,0 +1,387 @@
...
+        segs = _segments(seed=13, n_segments=2)
+        a = self._run(segs, use_inplace=False)
+        b = self._run(segs, use_inplace=True, start_positions=[1, 1])
+        _assert_states_equal(self, a, b)
+
+    def test_tail_alloc_branchless_flag_set_semantics(self):
```
Comment (sshleifer) [inline test/registered/unit/mem_cache/test_sync_free_swa_inplace.py:167]: Scenario: the flag is what `alloc_extend_swa_tail` sets when it installs zero mappings; with it set, even a fully-mapped segment takes the legacy path and states stay identical.

File: test/registered/unit/mem_cache/test_sync_free_swa_inplace.py [hunk 13]
```diff
@@ -0,0 +1,387 @@
...
+        return allocator
+
+    @staticmethod
+    def _swa_state(allocator):
+        sa = allocator.swa_attn_allocator
+        return (
+            sa.virtual_to_physical.clone(),
+            sa.physical_to_virtual.clone(),
+            sa._free_phys_pages.clone(),
+            sa.free_virtual_ids.clone() if sa.free_virtual_ids is not None else None,
```
Comment (sshleifer) [inline test/registered/unit/mem_cache/test_sync_free_swa_inplace.py:324]: `free_virtual_ids` is None on the non-id-owner swa side, hence the guard.

File: test/registered/unit/mem_cache/test_sync_free_swa_inplace.py [hunk 14]
```diff
@@ -0,0 +1,387 @@
...
+                continue
+            self.assertTrue(torch.equal(ta, tb), f"{name} differs:\n{ta}\n{tb}")
+        self.assertEqual(a[4], b[4], "live_page_count differs")
+
+    def test_unified_ps1_inplace_equals_free_swa(self):
+        def run(inplace):
+            allocator = self._build_composite(page_size=1, lazy=True)
+            v = allocator.alloc(8)
+            self.assertIsNotNone(v)
+            seg = v[2:6]
```
Comment (sshleifer) [inline test/registered/unit/mem_cache/test_sync_free_swa_inplace.py:343]: `v[2:6]` is a contiguous sub-range that is entirely swa-live.

File: test/registered/unit/mem_cache/test_sync_free_swa_inplace.py [hunk 15]
```diff
@@ -0,0 +1,387 @@
...
+            if inplace:
+                allocator.free_swa_segment_inplace(seg, start_pos=2)
+            else:
+                allocator.free_swa(seg)
+            return self._swa_state(allocator)
+
+        self._assert_swa_state_equal(run(False), run(True))
+
+    def test_unified_eager_compaction_falls_back(self):
```
Comment (sshleifer) [inline test/registered/unit/mem_cache/test_sync_free_swa_inplace.py:352]: Scenario: eager compaction has no sync-free formulation; the in-place call must route to `free_swa` and end in the same state.

File: test/registered/unit/mem_cache/test_sync_free_swa_inplace.py [hunk 16]
```diff
@@ -0,0 +1,387 @@
...
+                sa.physical_to_virtual.clone(),
+                sa.allocated_count(),
+            )
+
+        a, b = run(False), run(True)
+        for ta, tb in zip(a[:2], b[:2]):
+            self.assertTrue(torch.equal(ta, tb))
+        self.assertEqual(a[2], b[2])
+
+    def test_multi_ended_free_v_pages_lazy_twin_parity_paged(self):
```
Comment (sshleifer) [inline test/registered/unit/mem_cache/test_sync_free_swa_inplace.py]: Scenario (ps>1): `free(v_tokens)` runs `unique(v // ps)` (sorted); the twin takes the sorted distinct page ids, and states must match bitwise. Inside `run`: `alloc_with_virtual(v_pages)` binds 3 virtual pages on the swa side the way `alloc_extend` would, and `order = [2, 0, 1]` visits the pages deliberately out of ascending order to exercise the sort.

File: python/sglang/srt/mem_cache/allocator/paged.py [hunk 17]
```diff
@@ -302,6 +302,21 @@ def free_segment(self, free_index: torch.Tensor, *, start_pos: int):
                 self._copy_for_free_group(piece) for piece in pieces
             )
 
+    def free_page_ids(self, page_ids: torch.Tensor):
+        """Free whole pages by page id — the fixed-shape tail of free() for
```
Comment (alphabetc1) [inline python/sglang/srt/mem_cache/allocator/paged.py:306]: Too many comments. Can we simplify this to just one or two lines?

File: test/registered/unit/mem_cache/test_sync_free_swa_inplace.py [hunk 18]
```diff
@@ -0,0 +1,383 @@
+"""Unit tests for the fixed-shape in-place
```
Comment (alphabetc1) [inline test/registered/unit/mem_cache/test_sync_free_swa_inplace.py:1]: This test does not prove that the sync-free path is reached; please add explicit fast/fallback dispatch assertions.

File: test/registered/unit/mem_cache/test_sync_free_swa_inplace.py [hunk 19]
```diff
@@ -0,0 +1,383 @@
...
+            if inplace:
+                allocator.free_swa_segment_inplace(seg, start_pos=4)
+            else:
+                allocator.free_swa(seg)
+            return self._swa_state(allocator)
+
+        self._assert_swa_state_equal(run(False), run(True))
+
+
+if __name__ == "__main__":
```
Comment (alphabetc1) [inline test/registered/unit/mem_cache/test_sync_free_swa_inplace.py:391]: Several test cases repeat the same setup, legacy-vs-inplace execution, and state comparison. Could we consolidate them into a table-driven test to make the behavior matrix easier to maintain?

File: python/sglang/srt/mem_cache/allocator/swa.py (hunk 2 above)
Comment (alphabetc1) [inline python/sglang/srt/mem_cache/allocator/swa.py:289]: `_swa_mapping_may_be_partial` is allocator-wide and sticky, so once triggered, all subsequent requests fall back until the allocator is cleared. Is this intentional? Could it be tracked more locally or reset when no partial mappings remain?

**Comment by hnyls2002** [comment]
Closing as superseded by #38159 (merged), which lands the same mechanism on `main`:

- `free_swa_segment(idx, *, start_pos)` takes page representatives by `free_index[::page_size]`, gathers the mapping once at fixed shape, and frees the SWA side by page ids; no `torch.unique` and no `> 0` filter on the segment path.
- `free_page_ids(page_ids)` on the paged / token allocators is the same inner primitive; inside a free group the page ids are queued and flushed once.
- The partial-mapping guard is replaced by the `_SWA_PEER_MAPPED` invariant (a caller that wants the full side only calls `free_full_segment`), so every kv-row caller takes the fixed-shape path: window eviction, `SWARadixCache` node frees, and the unified `FreeComponentDeviceSlot` action, not only `free_swa_out_of_window_slots`.
- The unified allocator already had `free_swa` with `start_pos` (split into `free_swa_segment` in #38159), covering the multi-ended twin.

Thanks for the RL-workload measurement in the description; it is the clearest end-to-end number we have for this sync so far.
